numpy/core/src/multiarray/item_selection.c

```diff
@@ -787,255 +787,303 @@ PyArray_Choose(PyArrayObject *ip, PyObject *op, PyArrayObject *out,
-    if (rit == NULL || it == NULL) {
+    rit = (PyArrayIterObject *)PyArray_IterAllButAxis((PyObject *)rop, &axis);
+    if (it == NULL || rit == NULL) {
         goto fail;
     }
-    swap = !PyArray_ISNOTSWAPPED(op);
+    size = it->size;
 
     NPY_BEGIN_THREADS_DESCR(PyArray_DESCR(op));
-    if (argpart == NULL) {
-        argsort = PyArray_DESCR(op)->f->argsort[swhich];
-    }
-    else {
-        argpart = get_argpartition_func(PyArray_TYPE(op), pwhich);
-    }
-    size = it->size;
-    N = PyArray_DIMS(op)[axis];
-    elsize = PyArray_DESCR(op)->elsize;
-    astride = PyArray_STRIDES(op)[axis];
-    rstride = PyArray_STRIDE(ret,axis);
-
-    needcopy = swap || !(PyArray_FLAGS(op) & NPY_ARRAY_ALIGNED) ||
-                         (astride != (npy_intp) elsize) ||
-            (rstride != sizeof(npy_intp));
-    if (needcopy) {
-        char *valbuffer, *indbuffer;
 
-        valbuffer = PyDataMem_NEW(N*elsize);
+    if (needcopy) {
+        valbuffer = PyDataMem_NEW(N * elsize);
         if (valbuffer == NULL) {
             goto fail;
         }
-        indbuffer = PyDataMem_NEW(N*sizeof(npy_intp));
-        if (indbuffer == NULL) {
-            PyDataMem_FREE(valbuffer);
+    }
+
+    if (needidxbuffer) {
+        idxbuffer = (npy_intp *)PyDataMem_NEW(N * sizeof(npy_intp));
+        if (idxbuffer == NULL) {
             goto fail;
         }
-        while (size--) {
-            _unaligned_strided_byte_copy(valbuffer, (npy_intp) elsize, it->dataptr,
-                                         astride, N, elsize);
-            if (swap) {
-                _strided_byte_swap(valbuffer, (npy_intp) elsize, N, elsize);
-            }
-            iptr = (npy_intp *)indbuffer;
-            for (i = 0; i < N; i++) {
-                *iptr++ = i;
-            }
-            if (argpart == NULL) {
-                if (argsort(valbuffer, (npy_intp *)indbuffer, N, op) < 0) {
-                    PyDataMem_FREE(valbuffer);
-                    PyDataMem_FREE(indbuffer);
-                    goto fail;
+    }
+
+    while (size--) {
+        char *valptr = it->dataptr;
+        npy_intp *idxptr = (npy_intp *)rit->dataptr;
+        npy_intp *iptr, i;
+
+        if (needcopy) {
+            if (hasrefs) {
+                /*
+                 * For dtype's with objects, copyswapn Py_XINCREF's src
+                 * and Py_XDECREF's dst. This would crash if called on
+                 * an unitialized valbuffer, or leak a reference to
+                 * each object item if initialized.
+                 *
+                 * So, first do the copy with no refcounting...
+                 */
+                 _unaligned_strided_byte_copy(valbuffer, elsize,
+                                              it->dataptr, astride, N, elsize);
+                /* ...then swap in-place if needed */
+                if (swap) {
+                    copyswapn(valbuffer, elsize, NULL, 0, N, swap, op);
```

Comment:
I thought of it, but it would be a function with 9 or 10 parameters, unless we want to extract `hasrefs`, `copyswapn`, `elsize` and `swap` from `op` in each iteration.